### PR TITLE
[Rdy] [CR] Solar backpacks

### LINKED ISSUE #24010: Solar backpack concept
Since #23317 has put solar panels CBM on the road to oblivion, let's discuss @kevingranade idea of a solar backpack as a replacement.

Solar array can be folded like in satelites forming a special backpack. 
It would have no internal storage space other then space for the panels.
It would have (active) and (passive) state.
Passive state = folded panels hidden in bacpack's material. Backpack size, better encumbrance, better durability.
Active state = unfolded panels (butterfly wings style, see idea by @Night-Pryanik : #23317), big encumbrance, low durability

While active it could charge it's user via cable charger CBM or a new dedicated CBM, solar panel interface CBM (?) (somewhat like power armour interface CBM).
In second option this CBM might even come along with the backpack as a set.

It might have a quantum variant too.

Racionale: Backpack has all the pro's of former solar panels CBM without all its con's. Its wearable, its not covered by clothes, it can unfold so it would have decent surface for sun rays collection, it applies penalties from encubrance, weight, it has durability value that can be adjusted acordingly. All in all it would be lot of cons to live by with not so much pros, but it would be a nice replacement, as it would still be an option if one does not have any high end power CBM's at hand.

### PR DESCRIPTION
**Pull request by nexusmrsep**
Resolves #24010 

This PR introduces two variants of personal portable charging systems:
1. solar backpack (efficiency equal to former solar panels CBM),
2. quantum solar backpack (x2 efficiency).

**DONE**: items themselves, folding/unfolding actions, charging functions when worn
**TODO**: ~~spawning~~, ~~crafting recipes, uncrafting~~ [edit: done]

According to the initial concept solar pack needs to be worn and unfolded to work.
They are meant to work with cable charging system CBM, and are useless without it for self-charging.

They are fragile when unfolded and their encumbrance multiplies.
They are also quite heavy, and have no internal storage (frame holds only solar panels).

### CODE REVIEW

File: src/iuse.cpp [hunk 1]
```diff
@@ -3811,6 +3811,46 @@ int iuse::mp3_on(player *p, item *it, bool t, const tripoint &pos)
...
+    } else if ( !p->has_active_bionic("bio_cable") ) { // when OFF it takes no effect
+        p->add_msg_if_player(_( "Activate your cable charging system to take advantage of it." ));
+    }
+
+    if ( it->is_armor() && !( p->is_worn( *it ) ) ) {
+        p->add_msg_if_player( m_neutral, _( "You need to wear the %1$s before you can unfold it." ),
+                              it->tname().c_str() );
+    return 0;
+    }
+    p->add_msg_if_player(_( "You unfold solar array from the pack and plug it it." ));
```
Comment (Night-Pryanik) [inline src/iuse.cpp]: "plug it in".

**Comment by John-Candlebury** [comment]
Can you make the also charge standard UPSs in the player's inventory?

**Comment by nexusmrsep** [comment]
> If you dont mind taking feature requests, could you make them also charge standard UPSs in the player's inventory?

I don't mind feature requests. As for this particular one I have two concerns: one regarding the overall balance of the game and other about the approach.
Ad. 1 - Logic says that this might be an auxilary method to not only charge UPS but car batteries too. But would it be somehow too much influencial on game balance to allow it?
Ad. 2 -  UPS functions are awfuly nested and convoluted (for my skill level that is), and also profiled for either standard reload with batteries (consumed in the process) or via UPS recharging station - i'd need to find a good approach to pump some power directly from the solarpack. But its possible allright.

**Comment by John-Candlebury** [comment]
>Ad. 1 - Logic says that this might be an auxiliary method to not only charge UPS but car batteries too. But would it be somehow too much influential on game balance to allow it?

An UPS is explicitly an  universal power system, but a car battery could easily only take a certain voltage/current, so you can easily exclude them in those grounds.

 Don't think it would be a problem balance wise, the backpack is already unwieldy and heavy , it also generates power much more slowly than you consume it during fights or welding. I think it  would just add some flexibility and some Qol to non bionic builds.

Consider than Ideal game play for using laser weapons/power armor without bionics is basically dragging a shopping cart full of UPS's that you charged using a death-mobile,  and the backpack is not good enough to change this.

 What the backpack would do is almost negligibly delay the need of having the car when in combat, or let players use an ups for small utility stuff like daily cooking and the like.

**Comment by ZhilkinSerg** [comment]
I suggest changing material to `aluminum` - it is more obvious choice for some portable thing.

**Comment by DracoGriffin** [comment]
>Overall my opinion is: if aluminium was more common metal in-game then yes, otherwise no.

Find any extra-light frame vehicle and you'll get a lot of ingots. Likewise, if a player finds a heavy duty flashlight, that's a few more ingots. It is rare, but not terribly so.

**Comment by nexusmrsep** [comment]
How about I add new item: aluminium sheet, craftable from aluminium ingots. Then I switch metal sheets with aluminium sheets in recipes for solarpacks, and as a consequence their material wolud be aluminium. That could also be a method to popularize aluminium in game, as its rarity is kinda awkward. Someone could then verify some existing light metal items and change their composition and recipes for respectivly: aluminium, aluminium sheets and ingots. I like it from RP perspective, but Im not sure if aluminium is mellable enough to make sheets without having heavy factory equipment.

**Comment by RadHazard** [comment]
As an alternative source for aluminum plates, you could add lightweight quarterpanels made from aluminum on lightweight cars.  It's more work, but it would be a sensible source for premade aluminum plates.

[Based on a random metalworking forum I found](https://www.iforgeiron.com/topic/2981-forging-aluminum/), it seems like you should be able to cast or forge crude plates from aluminum by hand.

**Comment by nexusmrsep** [comment]
I agree on the malleability of aluminium, as i read a bit about it too. It is hard to solder, but not to weld. However this deeper "alluminium overhaul" would need separate PR, for now i only plan "aluminium sheets" on my TODO list. You may always join the effort.

**Comment by Night-Pryanik** [comment]
> However this deeper "alluminium overhaul" would need separate PR

In case anyone interested, there are #15369 and #16984 for reference.

**Comment by nexusmrsep** [comment]
After #24026 was merged Jenkins hates my iuse.cpp astyle even after running it throu Astyle. I have no idea if i can mitigate this situation on this end.

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

File: src/iuse.cpp [hunk 2]
```diff
@@ -3502,7 +3502,7 @@ int iuse::molotov_lit( player *p, item *it, bool t, const tripoint &pos )
         }
     } else {
         if( !t ) {
-            for( auto && pt : g->m.points_in_radius( pos, 1, 0 ) ) {
+            for( auto &&pt : g->m.points_in_radius( pos, 1, 0 ) ) {
```
Comment (ZhilkinSerg) [inline src/iuse.cpp]: My Astyle tells it should be `&& pt`.

**Comment by nexusmrsep** [comment]
@ZhilkinSerg yeah, whats strange that its either my Astyle own thign to propoce removing whitespece there or a random misclick on my side.. either way thanks for finding it. I couln't see it myself.

Also: Is there any good way to restart a failing job on Travis after making a pull request (and without pushing anything new)? I can restart whole build and single jobs of my master fork and subseqent branches from Travis' web page, but I cannot restart anything in the upstream (CleverRaven) even if I'm commit's author. All options are hidden for me. Travis fails are not due to commit errors, limited to single job/compiler and holds back many of my pull requests at bay.

**Comment by ZhilkinSerg** [comment]
It is possible to wear indefinite amount of solar backpacks at once (wear folded backpack, unfold it, wear another one and unfold it, repeat until satisfied).

image

**Comment by nexusmrsep** [comment]
Absolutly something to fix. As soon as I have some time.

File: src/bionics.cpp [hunk 3]
```diff
@@ -540,6 +540,13 @@ bool player::activate_bionic( int b, bool eff_only )
             add_msg_if_player( m_info,
                                _( "You need a jumper cable connected to a vehicle to drain power from it." ) );
         }
+        if( g->u.is_wearing( "solarpack_on" ) || g->u.is_wearing ( "q_solarpack_on" ) ) {
+            add_msg_if_player( m_info, _( "Your plugged-in solar pack is now able to charge"
```
Comment (ZhilkinSerg) [inline src/bionics.cpp]: Something strange here - why do you need line break between "charge" and "your"?

File: src/bionics.cpp [hunk 4]
```diff
@@ -540,6 +540,13 @@ bool player::activate_bionic( int b, bool eff_only )
...
+                " your system." ) );
+        } else if( g->u.is_wearing( "solarpack" ) || g->u.is_wearing ( "q_solarpack" ) ) {
+            add_msg_if_player( m_info, _( "You might plug in your solar pack to the cable charging"
```
Comment (ZhilkinSerg) [inline src/bionics.cpp]: Strange line break here too.

**Comment by nexusmrsep** [comment]
Ok, few changes after @ZhilkinSerg  found the multiplication bug (thanks for noticing!):
1. added flag for armor and tool_armor "ONLY_ONE" and a bypass function for the flag to ignore game constants for allowing wearing multiple items. Items with ONLY_ONE flag can be worn once.
2. Said flag went into both versions of solarpacks. I am aware it still allows wearing solarpack and quantum solarpack simultaneous.
3. I changed both pack activation function to disallow simultaneous unfolding of BOTH pack. When wearing both, you can activate ony ONE. Logic behind that: if game allows you to wear two backpacks, lets play along with that, but you still have ONE Cable Charging System CBM so only ONE can be plugged in at any time.
4. I also noticed that Cable C.S. CBM itself would use few interactions including detection of worn solar packs and their active/inactive state, so i added those too.

File: src/bionics.cpp (hunk 3 above)
Comment (nexusmrsep) [reply src/bionics.cpp]: 100 characters max length of line as stated in CODE_STYLE.md ?

**Comment by nexusmrsep** [comment]
Ok, so i basicly read throuout #15369 and #16984 and for my understanding structural aluminium is not a thing someone can make at home. Now I'm puzzled and asking for opinion: 

**1.** Should I realy add aluminium sheet PR and add it for example to helicopter crash site (I persume choppers want to be light hence aluminium), without adding any recipes to make it?
_**OR**_
**2.** Should I abandon this idea and leave metal sheets in solarpacks component list?

Adding new item just to use as one other item's component seems an overkill, so i need to know if it is rational - for example for further expansions. Please let me know.

**Comment by kevingranade** [comment]
Let's go with regular sheet metal.
I'd almost say it wants superalloy, but not for this PR.

**Comment by nexusmrsep** [comment]
Ok, then this PR is ready on my side. The UPS recharging option I'm going to queue on my todo list in a separate PR.

**Pull request merged**